### DISCUSSION TITLE
Generic function as input parameter of non-generic method

### DISCUSSION THREAD

**Discussion by gv-collibris** [dotnet/csharplang#5348] [Language Ideas]
In the following code:
```
public abstract class AbstractChildClass {}

public class ChildClass1 {}

public class ChildClass2 {}

public class ParentClass {
    public ChildClass1 Child1 { get; set; }
    public ChildClass2 Child2 { get; set; }

    public void ApplyActionToAllChildren(Action<T> action) where T : AbstractChildClass
    {
        action(Child1);
        action(Child2);
    }
}
```
the compiler fails on `T` not being defined in method `ApplyActionToAllChildren`. I have not found a work-around for this yet.
However, the code of the method is well-typed and all : you can just copy-paste these two lines where you define your `action`, and everything will be just fine.

DRYness cannot be respected here

**Comment by yaakov-h** [comment]
Why can't `ApplyActionToAllChildren` itself be generic? `T` has to be known for the action anyway...

**Reply by gv-collibris** [reply]
Because `T` depends on the child being handled.

usage example:
```
public class Helper
{
    private DbContext _context;
    public void Method<T>(T child) where T : AbstractChildClass
    {
        _context.Set<T>().Add(child);
    }

    public void ApplyMethodOnAllChildren(ParentClass parent)
    {
        parent.ApplyActionToAllChildren(Method);
    }
}
```
There, you actually need to have the right `T`, and not one of its inherited classes.
Also, having ` public void ApplyActionToAllChildren<T>(Action<T> action) where T : AbstractChildClass` tells you that `Child1` cannot be converted to `T`

it's actually logical : `T` has to be anything that inherits from `AbstractChildClass`, and `ChildClass1` is not "anything that inherits from `AbstractChildClass`"

**Reply by HaloFour** [reply]
The method itself must be generic in order to accept parameters that are generic.  That's how generics work in C# and in the .NET runtime.  If you don't want `ApplyActionToAllChildren` to be generic have it accept `Action<AbstractChildClass>` as the parameter type instead.

**Reply by gv-collibris** [reply]
does not work either as is: you have to cast `Child1` and `Child2` to `AbstractChildClass`.
The issue here is:
- `ApplyActionToAllChildren` cannot be generic
- `Method` has to be generic (because adding an abstract class in a database will not work)

The current solution is :
```
    public void ApplyMethodOnAllChildren(ParentClass parent)
    {
        Method(parent.Child1);
        Method(parent.Child2);
    }
```

If you have only a few children and 1 or 2 `Helper`s, it's ok. However, if you have ~10 children and ~10 `Helper`s then it just gets ugly...

**Reply by HaloFour** [reply]
> does not work either as is: you have to cast `Child1` and `Child2` to `AbstractChildClass`.

Upcasts happen automatically.

I really don't know what you're trying to write here but the way you're trying to apply generics doesn't make any sense.  Even if you could have the parameter be `Action<T>` you can't pass arbitrary instances of `AbstractChildClass` to it, that would violate type safety.  `Action<ChildClass2>` cannot accept a `ChildClass1`.

**Reply by gv-collibris** [reply]
I know that as of today, you cannot do what I want.

I'm more like wondering why not changing it slightly, what would be the issues with the use-case I have, etc.

The following code:
```cs
using System;

namespace Test
{
    class Program
    {
        static void Main(string[] args)
        {
            var parent = new ParentClass()
            {
                Child1 = new ChildClass1(),
                Child2 = new ChildClass2(),
            };
            Helper.ApplyMethodOnAllChildren(parent);
            Console.ReadLine();
        }
    }

    public abstract class AbstractChildClass { }

    public class ChildClass1 : AbstractChildClass { }

    public class ChildClass2 : AbstractChildClass { }

    public class ParentClass
    {
        public ChildClass1 Child1 { get; set; }
        public ChildClass2 Child2 { get; set; }

        public void ApplyActionToAllChildren(Action<AbstractChildClass> action)
        {
            action(Child1);
            action(Child2);
        }
    }

    public static class Helper
    {
        public static void Method<T>(T child) where T : AbstractChildClass
        {
            Console.WriteLine(typeof(T));
        }

        public static void ApplyMethodOnAllChildren(ParentClass parent)
        {
            parent.ApplyActionToAllChildren(Method);
        }
    }
}
```
will print :
```
Test.AbstractChildClass
Test.AbstractChildClass
```

That's all well and fine, but this is a simple case. Also, one might not want to see `AbstractChildClass`, but the actual type of the child.

This slightly more complex example:
```cs
using System;
using System.Collections.Generic;

namespace Test
{
    class Program
    {
        static void Main(string[] args)
        {
            var parent = new ParentClass()
            {
                Child1 = new List<ChildClass1>(),
                Child2 = new List<ChildClass2>(),
            };
            Helper.ApplyMethodOnAllChildren(parent);
            Console.ReadLine();
        }
    }

    public abstract class AbstractChildClass { }

    public class ChildClass1 : AbstractChildClass { }

    public class ChildClass2 : AbstractChildClass { }

    public class ParentClass
    {
        public ICollection<ChildClass1> Child1 { get; set; }
        public ICollection<ChildClass2> Child2 { get; set; }

        public void ApplyActionToAllChildren(Action<ICollection<AbstractChildClass>> action)
        {
            action(Child1);
            action(Child2);
        }
    }

    public static class Helper
    {
        public static void Method<T>(ICollection<T> child) where T : AbstractChildClass
        {
            Console.WriteLine(typeof(T));
        }

        public static void ApplyMethodOnAllChildren(ParentClass parent)
        {
            parent.ApplyActionToAllChildren(Method);
        }
    }
}
```
does not compile. You have to cast each child collection, like that:
```cs
        public void ApplyActionToAllChildren(Action<ICollection<AbstractChildClass>> action)
        {
            action(Child1 as ICollection<AbstractChildClass>);
            action(Child2 as ICollection<AbstractChildClass>);
        }
```
and you get the same result.

Note that if you write (in either of the both above examples):
```cs
        public static void ApplyMethodOnAllChildren(ParentClass parent)
        {
            //parent.ApplyActionToAllChildren(Method);
            Method(parent.Child1);
            Method(parent.Child2);
        }
```
you actually get 
```
Test.ChildClass1
Test.ChildClass2
```

Also, with:
```cs
     public class ParentClass
    {
        ...
        public void PrintAllChildrenTypes()
        {
            static void f<T>(ICollection<T> x) where T : AbstractChildClass => Console.WriteLine(typeof(T));
            f(Child1);
            f(Child2);
        }
    }
```
the `PrintAllChildrenTypes` method writes
```
Test.ChildClass1
Test.ChildClass2
```

The current limitation (from my perspective) is that the local function `f` cannot be passed as argument to the `PrintAllChildrenTypes` method while keeping the same behavior.

**Reply by Richiban** [reply]
The issue you're running into here is actually generic variance. You can't call `action(Child1);` because `ICollection<ChildClass1>` is not assignable to `ICollection<AbstractChildClass>`.

If you change all instances of `ICollection` to `IReadOnlyCollection` then your code will compile.

**Comment by OdinAnDieFreude** [comment]
Why even make it generic?

    public abstract class AbstractChildClass { }
    public class ChildClass1 : AbstractChildClass { }
    public class ChildClass2 : AbstractChildClass { }
    public class ParentClass
    {
      public ChildClass1 Child1 { get; set; }
      public ChildClass2 Child2 { get; set; }
      public void ApplyActionToAllChildren(Action<AbstractChildClass> action)
      {
        action(Child1);
        action(Child2);
      }
    }

With generics, you could pass ChildClass1 as the generic parameter, which means the action couldn't possibly run with a ChildClass2 object.

**Reply by gv-collibris** [reply]
see my comment above: https://github.com/dotnet/csharplang/discussions/5348#discussioncomment-1550843

also, with a slightly more complicated example:
```cs
using System;
using System.Collections.Generic;

namespace Test
{
    class Program
    {
        static void Main(string[] args)
        {
            var parent = new ParentClass()
            {
                Child1 = new List<ChildClass1>(),
                Child2 = new List<ChildClass2>(),
            };
            Helper.ApplyMethodOnAllChildren(parent);
            parent.PrintAllChildrenTypes();
            Console.ReadLine();
        }
    }

    public abstract class AbstractChildClass<TField>
    {
        public abstract TField GetField();
    }

    public class ChildClass1 : AbstractChildClass<string>
    {
        public override string GetField() => string.Empty;
    }

    public class ChildClass2 : AbstractChildClass<Guid>
    {
        public override Guid GetField() => Guid.Empty;
    }

    public class ParentClass
    {
        public ICollection<ChildClass1> Child1 { get; set; }
        public ICollection<ChildClass2> Child2 { get; set; }

        public void ApplyActionToAllChildren<U>(Action<ICollection<AbstractChildClass<U>>> action)
        {
            action(Child1 as ICollection<AbstractChildClass<U>>);
            action(Child2 as ICollection<AbstractChildClass<U>>);
            Console.WriteLine(Child1 as ICollection<AbstractChildClass<U>> == null); // if U != string, prints True
            Console.WriteLine(Child2 as ICollection<AbstractChildClass<U>> == null); // if U != Guid, prints True
        }

        public void PrintAllChildrenTypes()
        {
            void f<T, U>(ICollection<T> x) where T: AbstractChildClass<U>
                => Console.WriteLine($"{typeof(T)}, {typeof(U)}");
            f<ChildClass1, string>(Child1);
            f<ChildClass2, Guid>(Child2);
        }
    }

    public static class Helper
    {
        public static void Method<T, U>(ICollection<T> child) where T : AbstractChildClass<U>
        {
            Console.WriteLine($"{typeof(T)}, {typeof(U)}");
        }

        public static void ApplyMethodOnAllChildren(ParentClass parent)
        {
            parent.ApplyActionToAllChildren<string>(Method<AbstractChildClass<string>, string>); // I chose `string` arbitrarily, I could have put anything instead
        }
    }
}
```
Same as above, but you start having actual issues, like at least one of your children being passed to `action` as `null` in `ApplyActionToAllChildren`

**Reply by yaakov-h** [reply]
I feel like you're using `typeof(T)` where `obj.GetType()` is what you're really after.

**Reply by gv-collibris** [reply]
`typeof(T)` really is just an example. I had to find something illustrating the difference in behavior between `ApplyMethodOnAllChildren` and manually apply the same thing on each child.

**Reply by gv-collibris** [reply]
just replace `typeof(T)` with `child.Count` and `typeof(U)` with `string.Join(',', child.Select(c => c.GetField()))`, and brace yourself for NullRefs...

**Reply by Richiban** [reply]
It feels like perhaps what you're asking for in this second example is a function type (e.g. `Action`) that can represent a generic function?

**Reply by gv-collibris** [reply]
exactly

**Reply by Richiban** [reply]
Unfortunately that's not only a pretty huge feature (C# and the CLR would also need to be updated to allow for generic values†, not just methods), but of limited use: you'd only be use such a feature for input arguments to `Action` and `Func` since the method wouldn't have access to generic type parameters to be able to handle return values.

† So that this would have a type:
```cs
var f[T] = (x: T) => x;
```

**Reply by gv-collibris** [reply]
a local function does not have a type ?

**Reply by Richiban** [reply]
> a local function does not have a type ?

A local function is really a method, and methods can be generic. What I wrote is a lambda (which still has a type) but there's no such thing as a generic lambda.

**Reply by gv-collibris** [reply]
ok, I wasn't aware of local function being equivalent to methods.
I'm using them pretty much like any other variable, so I thought they were like lambdas...
